Validation/HLTrigger/plugins/HLTGenValClient.cc

```diff
@@ -0,0 +1,429 @@
+  TH1* efficHist = (TH1*) hDenominator->Clone(newEfficMEName.c_str());
+  efficHist->SetTitle(efficMETitle.c_str());
+  TClass* myHistClass = efficHist->IsA();
+  TString histClassName = myHistClass->GetName();
+  if (histClassName == "TH1F") {
+    efficME = ibooker.book1D(newEfficMEName, (TH1F*)efficHist);
+  } else if (histClassName == "TH2F") {
+    efficME = ibooker.book2D(newEfficMEName, (TH2F*)efficHist);
+  } else if (histClassName == "TH3F") {
+    efficME = ibooker.book3D(newEfficMEName, (TH3F*)efficHist);
+  }
+  delete efficHist;
+
+  // checking whether efficME was succesfully created
+  if (!efficME) {
+    edm::LogError("HLTGenValClient") << "computeEfficiency() : " << "Cannot book effic-ME from the DQM\n";
+    return;
+  }
+
+  // actually calculating the efficiency and filling the ME
+  genericEff(hDenominator, hNumerator, efficME);
+  efficME->setEntries(denominatorME->getEntries());
+
+  // Putting total efficiency in "GLobal efficiencies" histogram
+  if (makeGlobalEffPlot_) {
+
+    // getting global efficiency ME
+    HLTGenValClient::MonitorElement* globalEfficME = igetter.get(efficDir + "/globalEfficiencies");
+    if (!globalEfficME) // in case it does not exist yet, we create it
+      globalEfficME = ibooker.book1D("globalEfficiencies", "Global efficiencies", 1, 0, 1);
+    if (!globalEfficME) { // error handling in case creation failed
+      edm::LogError("HLTGenValClient") << "computeEfficiency() : " << "Cannot book globalEffic-ME from the DQM\n";
+      return;
+    }
+    globalEfficME->setEfficiencyFlag();
+
+    // extracting histogram
+    TH1F* hGlobalEffic = globalEfficME->getTH1F();
+    if (!hGlobalEffic) {
+      edm::LogError("HLTGenValClient") << "computeEfficiency() : "  << "Cannot create TH1F from ME, globalEfficME\n";
+      return;
+    }
+
+    // getting total counts
+    const float nDenominatorAll = hDenominator->GetEntries();
+    const float nNumeratorAll = hNumerator->GetEntries();
+
+    // calculating total efficiency
+    float efficAll = 0;
+    float errorAll = 0;
+    efficAll = nDenominatorAll ? nNumeratorAll / nDenominatorAll : 0;
+    errorAll = nDenominatorAll && efficAll < 1 ? sqrt(efficAll * (1 - efficAll) / nDenominatorAll) : 0;
+
+    // Filling the histogram bin
+    const int iBin = hGlobalEffic->Fill(newEfficMEName.c_str(), 0);
+    hGlobalEffic->SetBinContent(iBin, efficAll);
+    hGlobalEffic->SetBinError(iBin, errorAll);
+  }
+}
+
+// method to find all subdirectories of the given directory
+// goal is to fill myList with paths to all subdirectories
+void HLTGenValClient::findAllSubdirectories(DQMStore::IBooker& ibooker,
+                                             DQMStore::IGetter& igetter,
+                                             std::string dir,
+                                             std::set<std::string>* myList,
+                                             const TString& _pattern = TString("")) {
+  TString pattern = _pattern;
```

Comment:
in general in my experience it best to avoid as many root things as possible, ie things like TString and TPRegexp, its better to use the inbuild c++ ones

Still as its in now its fine for now